feat: add multiselect label managment to sample view

ReeceHoffmann opened the pull request:
**Changelog:**

Multiselect version of sample list view now permits label management for multiple samples at a time
 - Selecting a label will attempt to apply the label to all selected samples unless all select samples already have the label applied
 - selecting a label that is already applied to all selected labels will result in the label being cleared from all samples
 -  dropdown and label list indicate which labels are fully selected vs partly selected vs not selected. 
 - dropdown allows addition or removal of labels.

Images updated: 28/01/2022
Nothing selected:
![image](https://user-images.githubusercontent.com/59776400/151613209-3b736038-fb7c-474b-a2eb-46ee38af6bfd.png)


**Multiselect no dropdown:**
![image](https://user-images.githubusercontent.com/59776400/151613245-919c238d-7902-401c-9fd5-850e77438c7f.png)


**Multiselect dropdown:**
![image](https://user-images.githubusercontent.com/59776400/151613279-264e61da-335d-4a3f-a709-4b058a1441c4.png)

ReeceHoffmann committed:
feat: add multiselect label management to sample view

igboyes commented on src/js/samples/components/Sidebar/List.js:
```diff
@@ -20,3 +21,17 @@ export const SampleSidebarList = ({ items }) => {
 };
+
+export const SampleSidebarMultiselectList = ({ items }) => {
+    const sampleItemComponents = map(items, item => (
```
Unpack the object here:
```js
({ id, color, name, allLabeled }) => {}
```

ReeceHoffmann replied:
done

igboyes commented on src/js/samples/components/Sidebar/ManageLabels.js:
```diff
@@ -0,0 +1,120 @@
+
+const getSelectedSamples = (selected, documents) => {
+    return filter(documents, document => selected.includes(document.id));
+};
```
This is not efficient as you are calling `includes` for every `document`. This will bog down with larger sizes of `selected` and is a good example for thinking in terms of big O notation.

There is probably a function in Lodash that will do this for you much more efficiently.

ReeceHoffmann replied:
Turns out there was an existing `getSelectedSamples` in `samples/selectors.js` with identical implementation. I removed my own and replaced the code in the existing one with a version that should run faster.

igboyes reviewed (changes requested):
This is looking slick. I am noticing the following visual glitches:

* The shift in the heading of the labels sidebar component when shifting between the normal and selection context is a bit jarring. Can we eliminate that?

  It would be nice to do a universal fix that covers all sidebar components. I am fine considering a change in the pencil icon so that it takes up less vertical space.
* When changing from a check to nothing (all-selected to none-selected), the minus sign flashes for a split second.
* Hovering over the color dot on a label in the sidebar makes it turn black.

ReeceHoffmann commented:
Update 28/01/2022:

**Changes since last push:**
 - moved logic out of the `ManageLabels` component as much as possible
 - optimized the logic of `getSelectedSamples` and `getSelectedLabels`
 - fixed most known visual bugs 
 - added/updated tests as needed to accommodate the changes

igboyes reviewed (approved):
I am happy with this. Have a good weekend!

igboyes commented:
:tada: This PR is included in version 2.11.0 :tada:

The release is available on [GitHub release](https://github.com/virtool/virtool-ui/releases/tag/2.11.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
